sktime/dists_kernels/_base.py

```diff
@@ -0,0 +1,251 @@
+# -*- coding: utf-8 -*-
+"""
+Abstract base class for pairwise transformers (such as distance/kernel matrix makers)
+"""
+
+__author__ = ["fkiraly"]
+
+import numpy as np
+import pandas as pd
+
+from sktime.base import BaseEstimator
+
+from sktime.utils.validation.series import check_series
+
+
+class BasePairwiseTransformer(BaseEstimator):
+    """Base pairwise transformer for tabular or series data template class.
+
+    The base pairwise transformer specifies the methods and method
+    signatures that all pairwise transformers have to implement.
+
+    Specific implementations of these methods is deferred to concrete classes.
+    """
+
+    def __call__(self, X, X2=None):
+        """
+        Behaviour: returns pairwise distance/kernel matrix
+            between samples in X and X2
+                if X2 is not passed, is equal to X
+
+        alias for transform
+
+        Parameters
+        ----------
+        X: pd.DataFrame of length n, or 2D np.array with n rows
+        X2: pd.DataFrame of length m, or 2D np.array with m rows, optional
+            default X2 = X
+
+        Returns
+        -------
+        distmat: np.array of shape [n, m]
+            (i,j)-th entry contains distance/kernel between X.iloc[i] and X2.iloc[j]
+
+        Writes to self
+        --------------
+        symmetric: bool = True if X2 was not passed, False if X2 was passed
+            for use to make internal calculations efficient, e.g., in _transform
+        """
+        # no input checks or input logic here, these are done in transform
+        # this just defines __call__ as an alias for transform
+        return self.transform(X=X, X2=X2)
+
+    def transform(self, X, X2=None):
+        """
+        Behaviour: returns pairwise distance/kernel matrix
+            between samples in X and X2 (equal to X if not passed)
+
+        Parameters
+        ----------
+        X: pd.DataFrame of length n, or 2D np.array with n rows
+        X2: pd.DataFrame of length m, or 2D np.array with m rows, optional
+            default X2 = X
+
+        Returns
+        -------
+        distmat: np.array of shape [n, m]
+            (i,j)-th entry contains distance/kernel between X.iloc[i] and X2.iloc[j]
+
+        Writes to self
+        --------------
+        symmetric: bool = True if X2 was not passed, False if X2 was passed
```

Comment:
yes, I think that's a better convention.

Because `symmetric` should actually be a tag that encodes the property d(x,y) = d(y,x).
Then, if both X=X2 and d is symmetric, the algorithm for filling with i<j should be used.

Reply:
I've changed this now to the "correct" algorithm, including a renaming of the variable `X_equals_X2`.
That required introducing a tag - hope no one minds...